# Sustainable Development Report 2022 Data Exploration

[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/marialonsogar/fuzzy-compliance-SDG11/blob/main/fuzzy-sdg11/sd-report.ipynb)

# Table of contents
* [1. Introduction](#chapter1)
  * [1.1 Data source](#section1_1)
* [2. Requirements](#chapter2)
* [3. Exploration](#chapter3)
  * [3.1. 'About' Data](#chapter3_1)
  * [3.2. 'Codebook' Data](#chapter3_2)
  * [3.3. SDR2022 Data](#chapter3_3)
* [Summary](#summary)

# Introduction<a class="anchor" id="chapter1"></a>

The Sustainable Development Report assesses the progress of all 193 UN Member States on the SDGs

2. descripcion sdg11 
## motivation

## Data source <a class="anchor" id="section1_1"></a>

- The report, dashboard, country profiles and more related information can be downloaded and found at the following link: [https://dashboards.sdgindex.org/](https://dashboards.sdgindex.org/).
- The database of the Sustainable Development Report 2022 contains the SDG Index rankings, overall scores, spillover scores, dashboards and trends for all 17 SDGs and all 100+ indicators. The Codebook sheet lists indicator metadata and sources. The link to download data: [https://dashboards.sdgindex.org/static/downloads/files/SDR-2022-database.xlsx](https://dashboards.sdgindex.org/static/downloads/files/SDR-2022-database.xlsx)

# Requirements <a class="anchor" id="chapter2"></a>

In [1]:
import pandas as pd
import numpy as np
import itertools

# filter warnings
import warnings
warnings.filterwarnings('ignore')

# Exploration <a class="anchor" id="chapter3"></a>

The objective is to prepare a table with information related exclusively to SDG11.

In [2]:
# downloaded data saved in the repository
report_path = './data/raw-data/SDR-2022-database.xlsx'


In [3]:
# check sheet names in the excel data to read individual sheets later
full_data = pd.ExcelFile(report_path)
print(full_data.sheet_names)


['About', 'Overview', 'Codebook', 'SDR2022 Data', 'Raw Data', 'Raw Data - Trend Indicators', 'Backdated SDG Index']


## 'About' Data <a class="anchor" id="chapter3_1"></a>

The first step is to explore the data. We extract the relevant information from the 'About' sheet, which in particular describes the content of the other sheets.

In [4]:
# dataframe with the overall data info
about_df = pd.read_excel(full_data, 'About')

# indexs of relevant info (sheets descriptions and citation)
description_indexs = list(range(3, 8))
citation_indexs = 36

# print descriptions of the sheets
for i in description_indexs:
    print('*', about_df.iloc[i, 0], '-> ', str(about_df.iloc[i, 1]), '\n')


* Codebook ->  List of indicators, bounds and thresholds 

* SDR2022 Data ->  The SDR2022 data for all countries; includes spillover score, raw values, normalized scores, dashboard ratings, trends & goal scores. 

* Raw Data ->  Raw data for all countries by indicator 

* Raw Data - Trend inds ->  Panel data used for generating trend analysis 

* Backdated SDG Index ->  The SDG Index Score, and all goal and indicator scores, retroactively calculated across time using time series data that was carried forward in years with missing data. 



## 'Codebook' Data <a class="anchor" id="chapter3_2"></a>
Firstly, the Codebook sheet is checked by filtering by SDG 11 to find out which variables are currently used to measure it. From the data collected in this table, we will be particularly interested in the following fields: 
- `['IndCode', 'SDG', 'Indicator', 'Reference Year',
                     'Optimum (= 100)', 'Green threshold', 'Red threshold',
                     'Lower Bound (=0)', 'Justification for Optimum',
                     'Description', 'Imputation']`


In [5]:
codebook = pd.read_excel(full_data, 'Codebook')

# filter codebook by SDG==11
codebook_11 = codebook[codebook['SDG'] == 11]

# select columns
codebook_11 = codebook_11[['IndCode', 'SDG', 'Indicator', 'Reference Year',
                     'Optimum (= 100)', 'Green threshold', 'Red threshold',
                     'Lower Bound (=0)', 'Justification for Optimum',
                     'Description', 'Imputation']]
display(codebook_11)


IndCode  SDG                                          Indicator  \
74      sdg11_slums   11  Proportion of urban population living in slums...   
75       sdg11_pm25   11  Annual mean concentration of particulate matte...   
76   sdg11_pipedwat   11  Access to improved water source, piped (% of u...   
77  sdg11_transport   11             Satisfaction with public transport (%)   
78   sdg11_rentover   11                Population with rent overburden (%)   

    Reference Year  Optimum (= 100)  Green threshold  Red threshold  \
74            2018              0.0              5.0           25.0   
75            2019              6.3             10.0           25.0   
76            2020            100.0             98.0           75.0   
77            2021             82.6             72.0           43.0   
78            2019              4.6              7.0           17.0   

    Lower Bound (=0)   Justification for Optimum  \
74              90.0         Leave no one behind   
75              87.0  Average of best performers   
76               6.1         Leave no one behind   
77              21.0  Average of best performers   
78              25.6  Average of best performers   

                                          Description  \
74  Population living in slums is the proportion o...   
75  Air pollution measured as the population-weigh...   
76  The percentage of the urban population with ac...   
77  The percentage of the surveyed population that...   
78  Percentage of the population living in househo...   

                                           Imputation  
74  We assumed 0% population living in slums for h...  
75                                                NaN  
76                                                NaN  
77                                                NaN  
78                                                NaN

From the above, let us derive each indicator and its associated description.

In [6]:
# for each row of codebook_11, print the indicator and description
for i in range(len(codebook_11)):
    print('*', codebook_11.iloc[i, 2], '->', codebook_11.iloc[i, 9], '\n')

# print the citation
print(about_df.iloc[36, 1])

* Proportion of urban population living in slums (%) -> Population living in slums is the proportion of the urban population living in slum households. A slum household is defined as a group of individuals living under the same roof lacking one or more of the following conditions: access to improved water, access to improved sanitation, sufficient living area, housing durability, and security of tenur 

* Annual mean concentration of particulate matter of less than 2.5 microns in diameter (PM2.5) (μg/m³) -> Air pollution measured as the population-weighted mean annual concentration of PM2.5 for the urban population in a country. PM2.5 is suspended particles measuring less than 2.5 microns in aerodynamic diameter, which are capable of penetrating deep into the respiratory tract and can cause severe health damage. 

* Access to improved water source, piped (% of urban population) -> The percentage of the urban population with access to improved drinking water piped on premises. An "impro

In addition, from this table we can extract the ranges that will define the fuzzy variables defined later, as well as the minimum and maximum limits. Finally, the missing values of the first indicator are imputed as follows: 'We assumed 0% population living in slums for high-income OECD members for which no data was reported.'. No further imputation method is defined.

This table is saved below to './data/processed-data/' folder for further information. 

In [7]:
codebook_11.to_csv('./data/processed-data/sdg-variable-thresholds', index=False)

## SDR2022 Data <a class="anchor" id="chapter3_3"></a>

The SDR2022 data for all countries is studied now to get their indicators values, scores, etc. 

In [8]:
sdr = pd.read_excel(full_data, 'SDR2022 Data')

# select important columns
country_info = ['Country Code ISO3',
                'Country', 'Goal 11 Score', 'Goal 11 Dash']

# sdg indicators
sdg_11_indicators = ['Proportion of urban population living in slums (%)',
                     'Annual mean concentration of particulate matter of less than 2.5 microns in diam',
                     'Access to improved water source, piped (% of urban population)',
                     'Satisfaction with public transport (%)',
                     'Population with rent overburden (%)']

# list with column names of sdr if column contains 'sdg11'
sdg_11_info = [col for col in sdr.columns if 'sdg11' in col]
# remove from list above names containing one of ['Trend', 'Imputation', 'Year', 'Normalized', 'Dashboard']
sdg_11_info = [col for col in sdg_11_info if not any(x in col for x in ['Trend', 'Imputation', 'Year',
                                                                        'Normalized',
                                                                        'Dashboard'])]

# selected columns
country_info.extend(sdg_11_indicators)
country_info.extend(sdg_11_info)

# delete codes begining with _ (they are not countries)
sdr = sdr[~sdr['Country Code ISO3'].str.startswith('_')]

sdr = sdr[country_info]
sdr.sample(5, random_state=1)


Country Code ISO3           Country  Goal 11 Score Goal 11 Dash  \
44                DEU           Germany      90.908250       yellow   
69                GRD           Grenada            NaN         grey   
162               SVN          Slovenia      84.023333       yellow   
35                COD  Congo, Dem. Rep.      40.786000          red   
183               UZB        Uzbekistan      74.194000          red   

     Proportion of urban population living in slums (%)  \
44                                                 0.0    
69                                                 NaN    
162                                                0.0    
35                                                80.4    
183                                               58.5    

     Annual mean concentration of particulate matter of less than 2.5 microns in diam  \
44                                              11.281                                  
69                                              21.426                                  
162                                             15.372                                  
35                                              42.845                                  
183                                             26.904                                  

     Access to improved water source, piped (% of urban population)  \
44                                             100.000                
69                                                 NaN                
162                                                NaN                
35                                              67.411                
183                                             88.082                

     Satisfaction with public transport (%)  \
44                                     64.0   
69                                      NaN   
162                                    60.0   
35                                     41.0   
183                                    84.0   

     Population with rent overburden (%)  
44                                 4.154  
69                                   NaN  
162                                3.391  
35                                   NaN  
183                                  NaN

In [9]:
# save sdr to csv
sdr.to_csv('./data/processed-data/sdg_11.csv', index=False)


We need to know the missing values, how many countries have complete data and how many do not, and which variables they collect, etc. Below we define a table with the percentage of countries that have a combination of indicators available.

In [10]:
def table_percentage_countries_combination_indicators():
    """Table with the percentage of countries that have a value for each combination of indicators"""
    # define consistent notation with the article
    sdg_11_indicators_notation = [f'\\nu_{i}' for i in range(1, 6)]
    
    # create a dataframe with the combinations
    combinations_df = pd.DataFrame(columns=['Indicators', 'Percentage', 'Number of countries', 'Countries'])

    # create a list with all possible combinations of indicators
    for k in range(1, len(sdg_11_indicators)+1):
        for combination in itertools.combinations(sdg_11_indicators, k):
            # countries with combination not nan sdg_11_indicators such that the remaining indicators are nan 
            _nan_indicators_check = [indicator for indicator in sdg_11_indicators if indicator not in combination]
            countries_combination_data = sdr[(sdr[list(combination)].notna().all(axis=1)) & (sdr[_nan_indicators_check].isna().all(axis=1))]['Country']
            # number of countries with a combination
            number_of_countries = len(countries_combination_data)
            # compute percentage of countries with combination of indicators
            percentage_of_countries_with_combination = np.round(number_of_countries/len(sdr)*100, 2)
        
            if percentage_of_countries_with_combination != 0:
                # add to dataframe
                combinations_df = combinations_df.append({'Indicators': combination,
                                                            'Percentage': percentage_of_countries_with_combination,         
                                                            'Number of countries': number_of_countries,
                                                            'Countries': countries_combination_data}, 
                                                            ignore_index=True)
    # add a percentage symbol for each cell of col percentage
    combinations_df['Percentage'] = combinations_df['Percentage'].astype(str) + ' %'

    # map sdg_indicators_11 to the notation and replace each one of the Indicators df
    combinations_df['Indicators'] = combinations_df['Indicators'].apply(lambda x: [sdg_11_indicators_notation[sdg_11_indicators.index(indicator)] for indicator in x])
    combinations_df['Indicators'] = combinations_df['Indicators'].apply(lambda x: ', '.join(x))

    # add brakets to each Indicator set cell and use latex notation
    combinations_df['Indicators'] = combinations_df['Indicators'].apply(lambda x: f'${{{x}}}$')

    return combinations_df

comb = table_percentage_countries_combination_indicators()
display(comb)

## print the table as latex
# print(comb[['Indicators', 'Percentage', 'Number of countries']].to_latex(index=False))

Indicators Percentage Number of countries  \
0                              ${\nu_2}$     5.18 %                  10   
1                              ${\nu_3}$     0.52 %                   1   
2                       ${\nu_2, \nu_3}$     7.77 %                  15   
3                       ${\nu_2, \nu_4}$     3.63 %                   7   
4                ${\nu_1, \nu_2, \nu_3}$     3.11 %                   6   
5                ${\nu_1, \nu_2, \nu_4}$     1.04 %                   2   
6                ${\nu_2, \nu_3, \nu_4}$     9.33 %                  18   
7         ${\nu_1, \nu_2, \nu_3, \nu_4}$    51.81 %                 100   
8         ${\nu_1, \nu_2, \nu_4, \nu_5}$     2.59 %                   5   
9  ${\nu_1, \nu_2, \nu_3, \nu_4, \nu_5}$    14.51 %                  28   

                                           Countries  
0  7                 Antigua and Barbuda
18      ...  
1  112    Marshall Islands
Name: Country, dtype: ...  
2  3                Andorra
25     Brunei Darussa...  
3  4      United Arab Emirates
17                ...  
4  57                      Fiji
66             Gu...  
5  174    Trinidad and Tobago
185          Venezu...  
6  10             Azerbaijan
16               Bul...  
7  0       Afghanistan
1            Angola
2     ...  
8  9          Austria
83           Italy
86      ...  
9  8            Australia
12             Belgium
...

In order to subsequently define the fuzzy variables, we also need to know the universe of discourse of the variables. The universe of discourse of the variables that are measured in percentages will be [0, 100], by definition of percentage. We do not know that of the variable "Annual mean concentration of particulate matter of less than 2.5 microns in diameter (PM2.5) (μg/m³)". We will take as the lower limit of the universe of discourse the minimum value recorded for the indicator and similarly for the maximum. 

In [11]:
# print value of country with max value on 'Annual mean concentration of particulate matter of less than 2.5 microns in diam'
v2 = 'Annual mean concentration of particulate matter of less than 2.5 microns in diam'

min_v2 = sdr[sdr[v2] == sdr[v2].min()][v2].values[0]
max_v2 = sdr[sdr[v2] == sdr[v2].max()][v2].values[0]
print(f'Universe of discourse v2: [{min_v2}, {max_v2}]')


Universe of discourse v2: [5.102, 106.121]


# Summary <a class="anchor" id="summary"></a>

The data have been explored for the information necessary to subsequently define the variables and the fuzzy logic inference system. 
In particular, we have:
- Extracted the possible combinations of variables available and concluded that **8 different models** can be defined depending on the input variables. 
- Seen that only 28 countries (14.51 %) have information for all the five variables. 
- Seen that 11 countries (5.7 %) have information only for one variable, so they are marked as grey (not enough information available).
- Conclude that most of the countries (79.79%) have missing data for at least one indicator, difficulting the modeling stage.
- Determined the universe of discourse [5.102, 106.121] of one of the variables ('Annual mean concentration of particulate matter of less than 2.5 microns in diam'), wich was not clearly defined. 

# Session info


In [12]:
import session_info
session_info.show(html=False)

-----
numpy               1.23.4
pandas              1.5.1
session_info        1.0.0
-----
IPython             8.6.0
jupyter_client      7.4.4
jupyter_core        4.11.2
-----
Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)]
Windows-10-10.0.22000-SP0
-----
Session information updated at 2022-11-11 14:00
